Source/NSTimeZone.m

```diff
@@ -2905,349 +2905,195 @@ - (NSString*) timeZoneName
-      buf = NSZoneMalloc(NSDefaultMallocZone(), i);
-      types = (TypeInfo*)buf;
-      buf += (n_types * sizeof(TypeInfo));
-      trans = (int64_t*)buf;
-      buf += (n_trans * sizeof(int64_t)); 
-      idxs = (unsigned char*)buf;
-
-      /* Read in transitions. */
-      if (1 == version)
-	{
-	  /* v1 with 32-bit transitions */
-	  for (i = 0; i < n_trans; i++)
-	    {
-	      trans[i] = GSSwapBigI32ToHost(*(int32_t*)(bytes + pos));
-	      pos += sizeof(int32_t);
-	    }
-	}
-      else
-	{
-	  /* v2+ with 64-bit transitions */
-	  for (i = 0; i < n_trans; i++)
-	    {
-	      trans[i] = GSSwapBigI64ToHost(*(int64_t*)(bytes + pos));
-	      pos += sizeof(int64_t);
-	    }
+		      format: @"failed to parse zone data"];
 	}
-      for (i = 0; i < n_trans; i++)
-	{
-	  idxs[i] = *(unsigned char*)(bytes + pos);
-	  pos++;
-	}
-      for (i = 0; i < n_types; i++)
-	{
-	  struct ttinfo	*ptr = (struct ttinfo*)(bytes + pos);
-          uint32_t      off;
-
-	  types[i].isdst = (ptr->isdst != 0 ? YES : NO);
-	  types[i].abbr_idx = ptr->abbr_idx;
-          memcpy(&off, ptr->offset, 4);
-	  types[i].offset = GSSwapBigI32ToHost(off);
-	  pos += sizeof(struct ttinfo);
-	}
-      abbr = (unsigned char*)(bytes + pos);
-      {
-	id		abbrevs[charcnt];
-	unsigned	count = 0;
-	unsigned	used = 0;
-
-	memset(abbrevs, '\0', sizeof(id)*charcnt);
-	for (i = 0; i < n_types; i++)
-	  {
-	    int	loc = types[i].abbr_idx;
-
-	    if (abbrevs[loc] == nil)
-	      {
-		abbrevs[loc]
-		  = [[NSString alloc] initWithUTF8String: (char*)abbr + loc];
-		count++;
-	      }
-	    types[i].abbreviation = abbrevs[loc];
-	  }
-	/*
-	 * Now we have created all the abbreviations, we put them in an
-	 * array for easy access later and easy deallocation if/when
-	 * the receiver is deallocated.
-	 */
-	i = charcnt;
-	while (i-- > count)
-	  {
-	    if (abbrevs[i] != nil)
-	      {
-		while (abbrevs[used] != nil)
-		  {
-		    used++;
-		  }
-		abbrevs[used] = abbrevs[i];
-		abbrevs[i] = nil;
-		if (++used >= count)
-		  {
-		    break;
-		  }
-	      }
-	  }
-	abbreviations = [[NSArray alloc] initWithObjects: abbrevs count: count];
-	while (count-- > 0)
-	  {
-	    RELEASE(abbrevs[count]);
-	  }
-      }
-
-      GS_MUTEX_LOCK(zone_mutex);
-      [zoneDictionary setObject: self forKey: timeZoneName];
-      GS_MUTEX_UNLOCK(zone_mutex);
+#endif
+      scrub_abbrs(sp);
     }
   NS_HANDLER
     {
+      if (lsp)
+	{
+	  free(lsp);
+	}
       DESTROY(self);
-      NSLog(@"Unable to obtain time zone `%@'... %@", name, localException);
+      NSLog(@"Unable to obtain time zone `%@'... %@",
+	name, localException);
       if ([localException name] != fileException)
 	{
 	  [localException raise];
 	}
+      return nil;
     }
   NS_ENDHANDLER
+
+  GS_MUTEX_LOCK(zone_mutex);
+  [zoneDictionary setObject: self forKey: timeZoneName];
+  GS_MUTEX_UNLOCK(zone_mutex);
   return self;
 }
 
 - (BOOL) isDaylightSavingTimeForDate: (NSDate*)aDate
```

Comment:
Is it worthwhile to have this method and a few others? The super implementation would call timeZoneDetailForDate: and the behaviour would be the same, what we save here is the creation of the NSTimeZoneDetail object.